src/utils/generator.js

```diff
@@ -307,15 +312,22 @@ export class Generator {
 
   serialize(context: SerializationContext) {
     for (let bodyEntry of this.body) {
-      let nodes = bodyEntry.args.map((boundArg, i) => context.serializeValue(boundArg));
-      context.emit(bodyEntry.buildNode(nodes, context));
-      let id = bodyEntry.declaresDerivedId;
-      if (id !== undefined) context.announceDeclaredDerivedId(id);
+      if (!bodyEntry.isPure || !bodyEntry.declaresDerivedId || !context.canOmit(bodyEntry.declaresDerivedId)) {
+        let nodes = bodyEntry.args.map((boundArg, i) => context.serializeValue(boundArg));
+        context.emit(bodyEntry.buildNode(nodes, context));
```

Comment:
Even if you can omit: still buildNode and announce, just don't call .emit(). That way, the invariant I pointed out above should still hold.